src/show.jl

```diff
@@ -1,49 +1,32 @@
-show(io::IO, c::Colorant) = get(io, :compact, false) ? _showcompact(io, c) : _show(io, c)
-show(io::IO, c::ColorantNormed) = show_normed(io, c)
 
-# Nonparametric types
-show_normed(io::IO, c::Gray24) = print(io, "Gray24(", gray(c), ')')
-show_normed(io::IO, c::RGB24)  = print(io, "RGB24(", red(c), ',', green(c), ',', blue(c), ')')
-show_normed(io::IO, c::ARGB32) = print(io, "ARGB32(", red(c), ',', green(c), ',', blue(c), ',', alpha(c), ')')
+function show(io::IO, c::Colorant)
+    show_colorant_string_with_eltype(io, typeof(c))
+    _show_components(io, c)
+end
 
-# FIXME: handle `Color` and `TransparentColor` correctly
-#       (e.g. `Color{T,4}` such as CMYK is different from `Transparent3`).
-for N = 1:4
-    component = N >= 3 ? (:comp1, :comp2, :comp3, :alpha) : (:comp1, :alpha)
-    printargs = Array{Any}(undef, 2, N)
-    for i = 1:N
-        printargs[1,i] = :(show(io, $(component[i])(c)))
-        chr = i < N ? ',' : ')'
-        printargs[2,i] = :(print(io, $chr))
-    end
-    @eval begin
-        function _show(io::IO, c::Colorant{T,$N}) where T
-            print(io, colorant_string(typeof(c)), "{", T, "}(")
-            $(printargs[:]...)
-        end
+# Special handling for Normed types: don't print the giant type name
+function show(io::IO, c::ColorantNormed)
+    show_colorant_string_with_eltype(io, typeof(c))
+    if typeof(c) <: Union{RGB24, ARGB32, Gray24} # isempty(typeof(c).parameters) # Nonparametric types
+        _show_components(io, c) # with trailing "N0f8" unless :compat=>true
+    else
+        _show_components(IOContext(io, :compact=>true), c)
     end
+end
+
+function _show_components(io::IO, c::ColorantN{N}) where N
+    comp_n = (comp1, comp2, comp3, comp4, comp5)
+    print(io, '(')
     for i = 1:N
-        printargs[1,i] = :(show(IOContext(io, :compact => true), $(component[i])(c)))
-    end
-    @eval begin
-        function _showcompact(io::IO, c::Colorant{T,$N}) where T
-            print(io, colorant_string(typeof(c)), "{", T, "}(")
-            $(printargs[:]...)
-        end
-        # Special handling for Normed types: don't print the giant type name
-        function show_normed(io::IO, c::Colorant{FixedPointNumbers.Normed{T,f},$N}) where {T,f}
-            print(io, colorant_string(typeof(c)), '{')
-            FixedPointNumbers.showtype(io, eltype(typeof(c)))
-            print(io, "}(")
-            $(printargs[:]...)
-        end
+        show(io, (comp_n[i])(c))
```

Comment:
Assuming you've checked that this infers properly, I am amazed at how far Julia's compiler has come! (It makes sense that it's possible, but still really cool.)

Reply:
Sorry to disappoint you, but this does not infer properly. :sweat_smile:
I have determined that the cost of runtime dispatch is negligible, since I did not anticipate the need to print a large number of colors.  However, I will think about the improvement.

Reply:
You should be able to fix it with lispy tuple programming, e..g,
```julia
show_foreach(io, c, comp_n...)
```
where
```julia
@inline show_foreach(io, c, f1, f2, f3...) = (show(io, f1(c)); print(io, ','); show_foreach(io, c, f2, f3...))
@inline show_foreach(io, c, f1) = (show(io, f1(c)); print(io, ')'))
```

Reply:
I tried some variations, but the specialization of `show` seems quite complicated. In general, specializations have little effect and most instances are unnecessarily. On the other hand, the specialization for `ARGB32`, for example, is clearly effective.

The current code is not pretty good because we are leaving the specialization to the compiler (not using `@nospecialize`). 
However, there are no fatal problems. I'll do some more study on `show`, but you can also merge this PR.